src/NuGet.Services.Metadata.Catalog.Monitoring/Model/HijackedServiceIndexResourceV3.cs

```diff
@@ -0,0 +1,157 @@
+        {
+            TrackUsedEntries(orderedTypes);
+            return RedirectUri(orderedTypes) ?? base.GetServiceEntryUri(orderedTypes);
+        }
+
+        public override IReadOnlyList<Uri> GetServiceEntryUris(params string[] orderedTypes)
+        {
+            TrackUsedEntries(orderedTypes);
+            return RedirectUriList(orderedTypes) ?? base.GetServiceEntryUris(orderedTypes);
+        }
+
+        public override IReadOnlyList<Uri> GetServiceEntryUris(NuGetVersion clientVersion, params string[] orderedTypes)
+        {
+            TrackUsedEntries(orderedTypes);
+            return RedirectUriList(orderedTypes) ?? base.GetServiceEntryUris(clientVersion, orderedTypes);
+        }
+        #endregion
+
+        #region Hijackable interface implementation
+        private bool _isCollectingUsedEntries = false;
+
+        private List<string> _usedTypePrefixes = new List<string>();
+
+        private IDictionary<string, ServiceIndexEntry> _redirectMapping = null;
+
+        public void BeginGetUsedEntryPrefixes()
+        {
+            _isCollectingUsedEntries = true;
+            _usedTypePrefixes = new List<string>();
+            _redirectMapping = null;
+        }
+
+        public IEnumerable<string> EndGetUsedEntryPrefixes()
+        {
+            _isCollectingUsedEntries = false;
+            var usedEntries = _usedTypePrefixes;
+            _usedTypePrefixes = new List<string>();
+            return usedEntries;
+        }
+
+        public IEnumerable<ServiceIndexEntry> GetServiceIndexEntriesByTypePrefix(string typePrefix)
+        {
+            return Entries.Where(e => GetTypePrefix(e.Type) == typePrefix);
+        }
+
+        public void SetRedirect(IDictionary<string, ServiceIndexEntry> redirectMapping)
+        {
+            _redirectMapping = redirectMapping;
+        }
+
+        private ServiceIndexEntry Redirect(params string[] orderedTypes)
+        {
+            var typePrefix = GetTypePrefix(orderedTypes.FirstOrDefault());
+            if (_redirectMapping != null && _redirectMapping.Keys.Contains(typePrefix))
+            {
+                return _redirectMapping[typePrefix];
+            }
+
+            return null;
+        }
+
+        private Uri RedirectUri(params string[] orderedTypes)
+        {
+            var redirectedEntry = Redirect(orderedTypes);
+            return redirectedEntry?.Uri;
+        }
+
+        private IReadOnlyList<ServiceIndexEntry> RedirectList(params string[] orderedTypes)
+        {
+            var redirectedEntry = Redirect(orderedTypes);
+            return redirectedEntry != null ? new ServiceIndexEntry[] { redirectedEntry } : null;
+        }
+
+        private IReadOnlyList<Uri> RedirectUriList(params string[] orderedTypes)
+        {
+            var redirectedUri = RedirectUri(orderedTypes);
+            return redirectedUri != null ? new Uri[] { redirectedUri } : null;
+        }
+
+        private void TrackUsedEntries(params string[] orderedTypes)
```

Comment:
>TrackUsedEntries [](start = 21, length = 16)

Please make this thread safe. This is tight integration with client code and we have no guarantee of single threadedness, right?